### ISSUE TITLE
Speed up write?

### ISSUE THREAD

**Issue by catap** [beetbox/beets#5029]
### Proposed solution

Speed up write speed via network.

Right now beets writes the same file twice (tested on `.flac` but probably it's true for all format). The first time it had happened when files is moved into library. It works as expected via temporarry name. Anyway, when it edits tags, it reuploads the whole file with edited tags.

I see this behaviour quite well when upload heavy `.flac` into library via WebDAV by investigating `.davfs.*` files in the distination folder via SSH to the NAS.

Rework editing of tags from save full file, into append or truncate and replacing only tag part at the end of the file.

**Comment by wisp3rwind** [comment]
> Right now beets writes the same file twice (tested on `.flac` but probably it's true for all format). The first time it had happened when files is moved into library. It works as expected via temporarry name. Anyway, when it edits tags, it reuploads the whole file with edited tags.

It is quite unclear to me which commands you're running here. Could you specify the actual beets commands where you run into this issue? I guess that would be required as a starting point to track down why exactly this happens.

**Comment by catap** [comment]
@wisp3rwind I simple import an album via *very* slow internet and started investigation how it works :)

**Comment by sampsyo** [comment]
Indeed; it would be great to know when/how these writes happen… is there any chance it's a consequence of your configuration or plugins you're using? Understanding a bit more about when the writes happen, and where they arise in the beets code, could help a lot.

**Comment by catap** [comment]
@sampsyo what information is required? list of plugins / configuration seems to be useless for now, because it means to check everything.

Maybe a kind of debug log which includes each possible write? Or... I may do it another way: copy file on NAS side and compare it O:-)

**Comment by sampsyo** [comment]
It's sort of an open-ended recommendation, to be honest! In general, to understand how to proceed here, we will need to understand where those two writes are happening on your machine. A few options could include:

* try disabling plugins successively until it goes away; the last one disabled was the culprit
* if disabling all plugins still exhibits the problem, similarly try turning off config options
* if the problem still happens with a "vanilla" import, yeah, maybe something like `strace` or similar could reveal the I/O behavior

**Comment by arogl** [comment]
I see similar issues while importing and moving files on the same drive.

It seems beet imports the files, into the DB as temporary files in current location, does some work on some tags and then moves the files, updates the DB, if replaygain enabled, now on the moved files run that, then update tags and database. I am not sure if the RG update is direct or via temporary files in the same folder.

I'll try to do more testing later and provide some output

**Comment by catap** [comment]
@arogl this is almost the same things that I feel / saw.

@sampsyo  / @wisp3rwind I'll try to investigate in this year

**Comment by catap** [comment]
@arogl as I pointed in the beginning: edit of tags means save the whole file, not just append at the end of it ;)

**Comment by catap** [comment]
Seems that this is behaviour of mutagen: https://github.com/quodlibet/mutagen/blob/main/mutagen/id3/_file.py#L223-L274

I feel a kind of dejavu. I have reported near the similar "issue" here or in mutagen or somewhere else BTW

**Comment by catap** [comment]
Yes, it is.

**Comment by wisp3rwind** [comment]
To summarize the above: This seems to be slow because mutagen needs to modify the files by inserting/removing data somewhere, which is usually not just at the end of files. That in turn will be slow if the file system doesn't support doing so on the remote side and emulates it by completely rewriting the file. It's limited what we can do on beets side to speed things up.

I'm going to remove the `needinfo` tag, but this still requires more research by someone to figure out what we can or cannot do to improve the situation. Below, I'm collecting a few ideas.

---

As mentioned by @arogl, at the cost of an additional local copy we could first write tags locally and then move to the NAS, at least during import in the file manipulation stage.

However, we'd probably need some heuristics to decide when to do this. (Cross-filesystem move might be an indication, but maybe not enough to warrant this.) We probably don't want to do this if the target file system is local, and sufficiently capable. I don't know how easy it is to get such information about the filesystem, someone should probably try to investigate this.

It also looks like the scrub plugin is doing extra writes during import: By running in the `import_task_files` hook, it will only act after files have been moved/copied/written.

What we might attempt to do is to change the `manipulate_files` stage to perform the tag write via mutagen completely in-memory, at least for `MoveOperation.MOVE`: That is, read the file to a RAM-backed temp file, do all tag manipulation, offer a new hook that the scrub plugin can use to also modify that in-memory file, and only then write out to the final location on disk.
